src/PrestaShopBundle/Form/Admin/Sell/Product/Shipping/ShippingType.php

```diff
@@ -110,18 +116,35 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'default_empty_data' => 0.0,
                 'modify_all_shops' => true,
             ])
-            ->add('carriers', ChoiceType::class, [
-                'choices' => $this->carrierChoiceProvider->getChoices(),
+            // related all shops checkbox is rendered in javascript side, but it is still needed here to be correctly filled with data
+            ->add($modifyAllShopsInputName, HiddenType::class)
+        ;
+
+        $builder->addEventListener(FormEvents::PRE_SET_DATA, function (FormEvent $event) use ($modifyAllShopsInputName) {
```

Comment:
I think you made this a bit too complicated, sorry I didn't understand the need when we discussed it.
All the data attributes you add here can be added without the an event form since most of them are not based on the data anyway.

The only which is related to the data is `data-selected-carrier-ids` but you don't need it, in your JS function `initCarrierSelector` you use it to get the selected carriers, but you can get them by getting the initial select values no need for JSON here, you can use a querySelector with the pseudo attribute `:checked` to retrieve this list.

Same thing for the whole `carrierChoices` list you can fetch it from all the available options in the select.

Reply:
yep thanks, I had this feeling it is getting too complicated 😄 . Now its much better 👍